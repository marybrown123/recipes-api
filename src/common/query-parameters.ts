import { ApiProperty } from '@nestjs/swagger';
import { Transform } from 'class-transformer';
import { IsNumber, IsOptional, IsString } from 'class-validator';

export class QueryParameters {
  @IsString()
  @IsOptional()
  @ApiProperty({ example: 'dumplings', type: 'string' })
  name?: string;

  @Transform(({ value }) => (value ? parseInt(value) : null))
  @IsNumber()
  @IsOptional()
  @ApiProperty({ example: 1, type: 'number' })
  page = 1;

  @Transform(({ value }) => (value ? parseInt(value) : null))
  @IsNumber()
  @IsOptional()
  @ApiProperty({ example: 3, type: 'number' })
  limit = 3;
}
