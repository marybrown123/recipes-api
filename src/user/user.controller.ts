import { Body, Controller, Post } from '@nestjs/common';
import * as bcrypt from 'bcrypt';
import { UserService } from './user.service';
import { CreateUserDTO } from './DTOs/create-user.DTO';
@Controller('/user')
export class UserController {
  constructor(private readonly usersService: UserService) {}

  @Post('/signup')
  async createUser(@Body() user: CreateUserDTO) {
    const hashedPassword = await bcrypt.hash(user.password, 10);
    const userForDb = {
      name: user.name,
      password: hashedPassword,
    };
    const createdUser = await this.usersService.createUser(userForDb);
    return createdUser;
  }
}
