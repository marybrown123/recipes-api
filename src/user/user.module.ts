import { Module } from '@nestjs/common';
import { UserService } from './user.service';
import { UserController } from './user.controller';
import { PrismaService } from '../prisma/prisma.service';
import { MailModule } from '../mail/mail.module';
import { TokenModule } from '../token/token.module';
import { WebhookModule } from '../webhook/webhook.module';

@Module({
  imports: [MailModule, TokenModule, WebhookModule],
  providers: [UserService, PrismaService],
  exports: [UserService],
  controllers: [UserController],
})
export class UserModule {}
