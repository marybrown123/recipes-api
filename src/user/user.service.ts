import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import { PrismaService } from 'src/prisma/prisma.service';
import { CreateUserDTO } from './DTOs/create-user.DTO';
import { Role } from '@prisma/client';
import { UserResponse } from './responses/user.response';
import * as bcrypt from 'bcrypt';

@Injectable()
export class UserService {
  constructor(private prisma: PrismaService) {}

  async createUser(user: CreateUserDTO) {
    const userFromDb = await this.prisma.user.findUnique({
      where: { name: user.name },
    });

    if (userFromDb) {
      throw new HttpException('User already exists', HttpStatus.BAD_REQUEST);
    }

    const hashedPassword = await this.hashPassword(user.password);

    const userForDb = await this.prisma.user.create({
      data: {
        name: user.name,
        password: hashedPassword,
        roles: [Role.USER],
      },
    });

    return new UserResponse(userForDb);
  }

  async findOne(name: string) {
    return await this.prisma.user.findUnique({
      where: {
        name,
      },
    });
  }

  async hashPassword(password: string) {
    const hashedPassword = await bcrypt.hash(password, 10);
    return hashedPassword;
  }

  async generateAdminAccount() {
    const adminName = process.env.ADMIN_NAME;
    const adminFromDb = await this.prisma.user.findUnique({
      where: { name: adminName },
    });

    if (adminFromDb) {
      return true;
    }

    const hashedAdminPassword = await this.hashPassword(
      process.env.ADMIN_PASSWORD,
    );

    return !!(await this.prisma.user.create({
      data: {
        name: adminName,
        password: hashedAdminPassword,
        roles: [Role.ADMIN],
      },
    }));
  }
}
