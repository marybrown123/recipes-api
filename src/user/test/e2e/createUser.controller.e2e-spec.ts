import { HttpStatus, INestApplication, ValidationPipe } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import { CreateUserDTO } from '../../../user/DTOs/create-user.DTO';
import { UserServiceMock } from '../../../user/test/mocks/user.service.mock';
import { UserService } from '../../../user/user.service';
import * as request from 'supertest';
import { UserModule } from '../../../user/user.module';

describe('User Controller - Create', () => {
  let app: INestApplication;
  const path = '/user/signup';

  const correctPayload: CreateUserDTO = {
    name: 'testName',
    password: 'testPassword',
  };

  const validationTestStructrue = [
    {
      property: 'name',
      value: 1,
      description: 'should throw validation error when name is number',
    },
    {
      property: 'password',
      value: 1,
      description: 'should throw validation error when password is number',
    },
    {
      property: 'password',
      value: 'short',
      description:
        'should throw validation error when password is shorter than 8 characters',
    },
  ];

  beforeAll(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [UserModule],
    })
      .overrideProvider(UserService)
      .useClass(UserServiceMock)
      .compile();

    app = moduleFixture.createNestApplication();
    app.useGlobalPipes(new ValidationPipe({ transform: true }));
    await app.init();
  });

  it('should create a user', async () => {
    return request(app.getHttpServer())
      .post(path)
      .send(correctPayload)
      .expect(HttpStatus.CREATED)
      .then((res) => {
        expect(res.body.name).toBe('testName');
        expect(res.body.password).toBe(undefined);
      });
  });

  validationTestStructrue.map((object) => {
    const uncorrectPayload = { ...correctPayload };
    uncorrectPayload[object.property] = object.value;
    return it(object.description, async () => {
      return request(app.getHttpServer())
        .post(path)
        .send(uncorrectPayload)
        .expect(HttpStatus.BAD_REQUEST)
        .then((res) => {
          expect(res.body).toMatchSnapshot();
        });
    });
  });
});
