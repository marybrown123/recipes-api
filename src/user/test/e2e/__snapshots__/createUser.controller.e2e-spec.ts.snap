// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Controller - Create should throw validation error when name is number 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "name must be a string",
  ],
  "statusCode": 400,
}
`;

exports[`User Controller - Create should throw validation error when password is number 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "password must be longer than or equal to 8 characters",
    "password must be a string",
  ],
  "statusCode": 400,
}
`;

exports[`User Controller - Create should throw validation error when password is shorter than 8 characters 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "password must be longer than or equal to 8 characters",
  ],
  "statusCode": 400,
}
`;
